internal/pkg/bashexec/bashexec.go

```diff
@@ -0,0 +1,60 @@
+package bashexec
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+// BashRunner is a wrapper for executing in-memory bash scripts
+type BashRunner struct {
+	name string
+	cmd  *exec.Cmd
+}
+
+// NewBashRunner creates a bash executor from in-memory shell script.
```

Comment:
It's not immediately clear to me (I feel like I should know this but don't) why we need to convert the shell script (file) to be in memory first. What benefits does this have? Should we add a comment/explanation?

Reply:
We now have 3 different ways to interact with shell.

- `pkg/internal/cosash` is a "co-processing" proxy for `cmdlib.sh`
- `pkg/internal/bashexec` (what you're commenting on) is about a convenience wrapper for executing shell script embedded in memory in Go (which may or may not use `cmdlib.sh`, but mostly for things that aren't)
- Running fully external scripts as a subprocess, things that usually live in `/usr/lib/coreos-assembler/foo.sh`

The rationale for in-memory embedding is to aid *incremental* porting of things like `cmd-build.sh`.  Basically we might start by changing the option processing to be Go, but use Go file embedding to actually move the data from living in `/usr/lib/coreos-assembler` to being in the binary.

This makes it clearer that the goal is to use Go, not to have distinct shell scripts.

Reply:
> We now have 3 different ways to interact with shell.
> 
> * `pkg/internal/cosash` is a "co-processing" proxy for `cmdlib.sh`

I feel like I kind of understand this but it's a struggle. This is brilliant and helps us get closer to our goal but I think we're starting to violate a [basic principle](https://www.goodreads.com/quotes/273375-everyone-knows-that-debugging-is-twice-as-hard-as-writing) here:

```
“Everyone knows that debugging is twice as hard as writing a
program in the first place. So if you're as clever as you
can be when you write it, how will you ever debug it?”
```

IOW, how many people on our team fully grok what is going on here and how many people can debug it?


> * `pkg/internal/bashexec` (what you're commenting on) is about a convenience wrapper for executing shell script embedded in memory in Go (which may or may not use `cmdlib.sh`, but mostly for things that aren't)

I guess I still don't really understand the "embedded in memory" part.  Why can't we use a file on disk rather than the `fd/3` magic. I'm missing something fundamental about why that part is required. Maybe the new process doesn't have access to the old filesystem view of the world?

> * Running fully external scripts as a subprocess, things that usually live in `/usr/lib/coreos-assembler/foo.sh`

I feel like this is the easiest option to understand, I was even able to figure it out myself on Friday, but obviously we aren't able to take as "incremental" of an approach with this strategy.

> 
> 
> The rationale for in-memory embedding is to aid _incremental_ porting of things like `cmd-build.sh`. Basically we might start by changing the option processing to be Go, but use Go file embedding to actually move the data from living in `/usr/lib/coreos-assembler` to being in the binary.
> 
> This makes it clearer that the goal is to use Go, not to have distinct shell scripts.

Reply:
> I feel like I kind of understand this but it's a struggle. This is brilliant and helps us get closer to our goal 

Hah, thanks - I appreciate the complement, seriously!  I'll be honest I was quite happy when I typed out that code and it all basically worked the first time I ran it.  But, it's probably because I have a *lot* of experience with gluing together code in multiple languages and have learned the hard way what works and what doesn't.  (And really more generally, things like [what can work but is overkill and hard to debug](https://en.wikipedia.org/wiki/Common_Object_Request_Broker_Architecture))

e.g. I also have worked on [gobject introspection](https://gi.readthedocs.io/en/latest/) and https://github.com/containers/containers-image-proxy-rs 

> but I think we're starting to violate a [basic principle](https://www.goodreads.com/quotes/273375-everyone-knows-that-debugging-is-twice-as-hard-as-writing) here:

I think this is a valid point.  But:

- I personally can't think of a simpler approach that will also allow us to avoid duplicating or reimplementing the shell script.  The first path would lead to a lot of extra typing and new bugs, and the latter would easily scope-creep into a wholesale rewrite that never works (and would be a continual conflict fest until it merged)
- But on the flip side, I think with *some* rewriting in the near term future we can avoid the need for the special "cosash" co-processing path and just fork off "stateless" shell script, so this code *will* go away

Reply:
> > but I think we're starting to violate a [basic principle](https://www.goodreads.com/quotes/273375-everyone-knows-that-debugging-is-twice-as-hard-as-writing) here:
> 
> I think this is a valid point. But:
> 
> * I personally can't think of a simpler approach that will also allow us to avoid duplicating or reimplementing the shell script.  The first path would lead to a lot of extra typing and new bugs, and the latter would easily scope-creep into a wholesale rewrite that never works (and would be a continual conflict fest until it merged)

Right. I think we probably do end up doing this, but an alternative is that we go with the 'duplicating' approach with a plan and a timeline for migrating everything else. The problem here is I don't think we can are in a position to commit to that.

Sorry for the noise here.

> * But on the flip side, I think with _some_ rewriting in the near term future we can avoid the need for the special "cosash" co-processing path and just fork off "stateless" shell script, so this code _will_ go away

👍